### PR TITLE
Implementation of simple dimension dependent particle stepping

### LINKED ISSUE #4: Particle Tracking with 3D velocity propagating in lower dimension.
## Reason
PIC codes require we track all 3 velocity components for a given particle but only propagate the particles in the same dimension as the mesh has. 

## Design
- A new RayStudy which sets up particles to be tracked and resets them on each tilmestep 
- A RayKernel which sets the maximum distance and direction for the particle in the `preTrace` method 
- Velocity of particles will be tracked with `RayData` 
- `maxDistance` will be set based on the particle velocity, time step and the dimension of the mesh 

## Impact
Implements a fundamental requirement for PIC modeling

### PR DESCRIPTION
**Pull request by gsgall**
closes #4

### CODE REVIEW

**Comment by loganharbour** [comment]
You don't need to update the moose submodule, it's done automatically from our continuous integration suite at civet.inl.gov.

In general, you'll just want to rebase on the upstream (idaholab) remote of salamander to pull these changes in. If you setup the repo as we have it in the contributing docs (see https://github.com/idaholab/salamander/blob/devel/doc/content/sqa/contributing.md), the way to pull in the new content (i.e., moose submodule) would be:

```
git fetch upstream
git rebase upsteam/devel
```

Glad to help with this more, I'm just not sure as to what your level of comfort is with git

**Comment by gsgall** [comment]
I would say I have moderate comfort with using git I have used it while pushing and pulling but rebasing and merging are still somewhat unfamiliar territory for me. 

I've just tried a to rebase my fork with 

```
git remote add upstream git@github.com:idaholab/salamander.git
git fetch upstream
```

And I got a permission denied message 

```git rebase upsteam/devel
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: invalid upstream 'upsteam/devel'
```

I thought I should have permission for this repo so I am not sure what the issue with this is.

**Comment by simopier** [comment]
@gsgall could you run

```
cd ~/projects/salamander
git remote -v 
```

and share what comes up?

**Comment by gsgall** [comment]
Sorry about the bumpy start. The issue was that I had updated the submodule on my branch and then pushed rather than just rebasing wih the salamander repo. I had to delete a few commits where that included the moose submodule update and then set up  the salamander repo as the upstream of my fork with the following commands.  

```
git remote add upstream git@github.com:idaholab/salamander.git
git remote set-url upstream https://github.com/idaholab/salamander.git
git fetch upstream                                               
git rebase upsteam/devel
```

@loganharbour helped me sort out the issues I was having over slack. Really appreciate the support!

File: include/raykernels/PICKernelBase.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+// Local Includes
```
Comment (loganharbour) [inline include/raykernels/PICKernelBase.h]:
```suggestion
```

Our standard typically dictates that first includes are the things you depend on, so this context isn't needed

File: include/userobjects/PICStudyBase.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  virtual void generateRays();
+  /** Ray Data for storing all components of velocity */
+  const RayDataIndex _v_x_index;
+  const RayDataIndex _v_y_index;
+  const RayDataIndex _v_z_index;
+  /** Keep a second counter so that we don't try and set the
+   * ray direction again when it is passed from one processor to another
+   */
+  const RayDataIndex _direction_set_index;
```
Comment (loganharbour) [inline include/userobjects/PICStudyBase.h]: All data members need to be protected. If you want to access them from outside of the class, create a method to access them. In general, we always avoid accessing member variables directly from outside of a class

File: include/userobjects/PICStudyBase.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  // The banked rays to be used on the next timestep (restartable)
+  std::vector<std::shared_ptr<Ray>> & _banked_rays;
+
+protected:
+  virtual void postExecuteStudy();
```
Comment (loganharbour) [inline include/userobjects/PICStudyBase.h]:
```suggestion
  virtual void postExecuteStudy() override;
```

I think?

File: src/raykernels/PICKernelBase.C [hunk 4]
```diff
@@ -0,0 +1,78 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+// MOOSE includes
```
Comment (loganharbour) [inline src/raykernels/PICKernelBase.C]:
```suggestion

```

File: include/userobjects/PICStudyBase.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ * where Rays have propagate a bit each time step
+ */
+class PICStudyBase : public RayTracingStudy
+{
+public:
+  PICStudyBase(const InputParameters & parameters);
+
+  static InputParameters validParams();
+
+  virtual void generateRays();
```
Comment (loganharbour) [inline include/userobjects/PICStudyBase.h]:
```suggestion
  virtual void generateRays() override;
```

File: src/raykernels/PICKernelBase.C [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+void
+PICKernelBase::preTrace()
+{
+  auto dt = _fe_problem.dt();
+  setDirectionAndMaxDistance(dt);
+}
+
+void
+PICKernelBase::setDirectionAndMaxDistance(const Real dt)
```
Comment (loganharbour) [inline src/raykernels/PICKernelBase.C]: What is the thought process for setting the velocity inside the kernel instead of systematically setting it from the study?

When we need to sample from the velocity field, that sampling will be done in the study and not the kernel. The kernel should act on things that happen along each segment of a trace, not things that happen just once.

File: src/raykernels/PICKernelBase.C [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+{
+  // lets make sure we don't keep trying to set the starting direction
+  // if it has already been set when the ray is passed between processors.
+  if (currentRay()->data()[_direction_set_index])
+    return;
+  currentRay()->data()[_direction_set_index] = true;
+
+  auto ray = currentRay();
```
Comment (loganharbour) [inline src/raykernels/PICKernelBase.C]: Because you're not taking this as a reference, it's constructing a new `std::shared_ptr<Ray>` (which points to the same Ray). This is actually not a cheap operation. It's what we call an atomic operation, which is thread safe. Such operations aren't necessary here.

```
auto & ray 
```

is more appropriate

File: src/userobjects/PICStudyBase.C [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  params.set<bool>("_use_ray_registration") = false;
+
+  return params;
+}
+
+PICStudyBase::PICStudyBase(const InputParameters & parameters)
+  : RayTracingStudy(parameters),
+    _v_x_index(registerRayData("v_x")),
+    _v_y_index(registerRayData("v_y")),
+    _v_z_index(registerRayData("v_z")),
```
Comment (loganharbour) [inline src/userobjects/PICStudyBase.C:39]: Do you always need to register y and z?

File: src/userobjects/PICStudyBase.C [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+void
+PICStudyBase::generateRays()
+{
+}
+
+void
+PICStudyBase::postExecuteStudy()
+{
+  _banked_rays = rayBank();
```
Comment (loganharbour) [inline src/userobjects/PICStudyBase.C:90]: Add a comment here as to what this does, it's not the most straight forward

File: test/src/userobjects/TestPICStudyBase.C [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  auto params = PICStudyBase::validParams();
+
+  params.addRequiredParam<std::vector<Point>>("start_points",
+                                              "The point(s) where the ray(s) start");
+  params.addRequiredParam<std::vector<Point>>(
+      "start_velocities",
+      "The direction(s) that the ray(s) start in (does not need to be normalized)");
+  // We're not going to use registration because we don't care to name our rays because
+  // we will have a lot of them
+  params.set<bool>("_use_ray_registration") = false;
```
Comment (loganharbour) [inline test/src/userobjects/TestPICStudyBase.C]:
```suggestion
```
Already set in the base class

File: src/raykernels/PICKernelBase.C [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    velocity(1) = v(1);
+
+  if (dim == 3)
+    velocity(2) = v(2);
+
+  const auto max_dist = std::sqrt(velocity * velocity) * dt;
+  const auto direction = velocity.unit();
+
+  ray->setStartingMaxDistance(max_dist);
+  ray->setStartingDirection(direction);
```
Comment (loganharbour) [inline src/raykernels/PICKernelBase.C]: Yeah - this kind of thing should be done at the study level

**Review by loganharbour — CHANGES_REQUESTED** [review]
Mostly minor changes except for the fundamental question of why the kernel is the way that it is

File: src/userobjects/PICStudyBase.C (hunk 8 above)
Comment (gsgall) [reply src/userobjects/PICStudyBase.C:39]: All velocity components do need to be tracked during simulation even in lower dimensions. I think storing this data in ray data is the most intuitive way to do this but we could also use a scheme like the following. 

- Always maintain a velocity magnitude in ray data
- In 3D no additional data is needed to store velocity data 
- In 2D store the z component as ray data 
- In 1D store the y, and z components as ray data 

But overall we do need to store all 3 components of velocity even when the geometry may be 1D. Let me know if you have a preference for storing all 3 components in ray data or in the manner described above.

File: src/raykernels/PICKernelBase.C (hunk 11 above)
Comment (gsgall) [reply src/raykernels/PICKernelBase.C]: The thinking I had for modifying the direction and max distance settings for a ray in a RayKernel rather than in the Study was to allow for more flexibility in the particle stepping methods and have those be independent from the generation of particles. The first example I was thinking about for this would be for electrostatic versus electro magnetic PIC simulations. The particle push algorithm looks a little different between the two and I think it will reduce code replication if we have the kernels control that pushing algorithm for that reason. 

I can see another way forward with having the RayStudy control the velocity updating and the direction setting but what I could see happening if we do this is that the number of RayStudies could get a bit out of hand if want to offer a few different options for particle initialization algorithms. If the RayStudy controls the particle velocity update then we would need at a minimum two version of each study. One that uses the electric field alone update and one that uses the electromagnetic update. This would problem would also be worse if we wanted to investigate another particle stepping algorithms than the two standards, Leap Frog and Boris.

File: include/raykernels/PICKernelBase.h [hunk 12]
```diff
@@ -0,0 +1,42 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
```
Comment (cticenhour) [inline include/raykernels/PICKernelBase.h]:
```suggestion
//* This file is part of FENIX: Fusion Energy Integrated Multiphys-X,
//* A multiphysics application for modeling plasma facing components
//* https://github.com/idaholab/fenix
//*
//* FENIX is powered by the MOOSE Framework
//* https://www.mooseframework.inl.gov
//*
//* Licensed under LGPL 2.1, please see LICENSE for details
//* https://www.gnu.org/licenses/lgpl-2.1.html
```

We should create a FENIX file header instead of using the MOOSE one, and then replace all instances of this header in the PR. I've made an attempt at one here. Any thoughts @simopier @loganharbour?

**Review by cticenhour — CHANGES_REQUESTED** [review]
@loganharbour has this well in hand, but I did notice one item I wanted to make sure I highlighted.

File: src/userobjects/PICStudyBase.C (hunk 8 above)
Comment (loganharbour) [reply src/userobjects/PICStudyBase.C:39]: Nah - no need to complicate it. I just wanted to make sure it wasn't easy to remove

File: include/raykernels/PICKernelBase.h (hunk 12 above)
Comment (loganharbour) [reply include/raykernels/PICKernelBase.h]: Fine by me. remove the redundant www.mooseframework.inl.gov -> mooseframework.inl.gov

File: src/raykernels/PICKernelBase.C (hunk 11 above)
Comment (loganharbour) [reply src/raykernels/PICKernelBase.C]: The kernel fundamentally as a system is meant for how the rays propagate, while the study handles everything else. If this is a bit brittle, that's fine it just means we need another way to compartmentalize. 

We can easily have other objects be an input to the study to define how the stepping is done. But it should be separate from the propagation part

Comment (gsgall) [reply src/raykernels/PICKernelBase.C]: That sounds good. I have an initial thought of how to implement this.  Let me know what you think about this 

Create a new user object which has the ability to sample fields at the locations of particles or obtain the average value of a field inside the element in which the ray is contained. This object could then be passed as an input to the pic ray study and have the ray study should call a method in which the rules for how to update the particle velocity within the study are defined.

File: include/raykernels/PICKernelBase.h (hunk 12 above)
Comment (simopier) [reply include/raykernels/PICKernelBase.h]: Sorry, I don't seem to get github notifications. 

Yes, it looks good to me.

File: src/raykernels/PICKernelBase.C (hunk 6 above)
Comment (gsgall) [reply src/raykernels/PICKernelBase.C]: This logic was moved to a new user object which will be passes an input parameter to the RayTracingStudy and supplies the velocity updating logic.

File: include/userobjects/PICStudyBase.h [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+class PICStudyBase : public RayTracingStudy
+{
+public:
+  PICStudyBase(const InputParameters & parameters);
+
+  static InputParameters validParams();
+
+  virtual void generateRays() override;
+  // The banked rays to be used on the next timestep (restartable)
+  std::vector<std::shared_ptr<Ray>> & _banked_rays;
```
Comment (loganharbour) [inline include/userobjects/PICStudyBase.h:28]: Should not have public data members

File: include/userobjects/PICStudyBase.h [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+protected:
+  virtual void postExecuteStudy() override;
+  /** Ray Data for storing all components of velocity */
+  const RayDataIndex _v_x_index;
+  const RayDataIndex _v_y_index;
+  const RayDataIndex _v_z_index;
```
Comment (loganharbour) [inline include/userobjects/PICStudyBase.h]:
```suggestion
  /// Ray data for storing velocity components
  ///@{
  const RayDataIndex _v_x_index;
  const RayDataIndex _v_y_index;
  const RayDataIndex _v_z_index;
  ///@}
```

When you wrap in `///@{  ...  ///@}`, they show up grouped in doxygen

File: include/userobjects/PICStudyBase.h [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+  /** Ray data fro storing the number of real particles each ray represents */
+  const RayDataIndex _weight_index;
+
+  // Whether or not we've generated rays yet (restartable)
+  bool & _has_generated;
+
+  // the velocity updater object which we will hold the rules for how our
+  // particles velocities are updated
+  const VelocityUpdaterBase & _velocity_updater;
```
Comment (loganharbour) [inline include/userobjects/PICStudyBase.h]:
```suggestion
  /// Whether or not we've generated rays yet (restartable)
  bool & _has_generated;

  /// the velocity updater object which we will hold the rules for how our
  /// particles velocities are updated
  const VelocityUpdaterBase & _velocity_updater;
```

We typically comment vars with `///` and methods with `/*`

File: include/userobjects/PICStudyBase.h [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+//* FENIX is powered by the MOOSE Framework
+//* https://www.mooseframework.inl.gov
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Function.h"
+#include "ClaimRays.h"
```
Comment (loganharbour) [inline include/userobjects/PICStudyBase.h]:
```suggestion
```

I think that these are unused

File: include/userobjects/PICStudyBase.h [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+#include "RayTracingStudy.h"
+#include "VelocityUpdaterBase.h"
```
Comment (loganharbour) [inline include/userobjects/PICStudyBase.h]:
```suggestion

class VelocityUpdaterBase;
```

You can remove this and only "forward declare" it. All the compiler needs to know in this is that it is a class. When you go acutally call methods on this object (in the source file), you include it there. We put forward declarations after an extra newline after the includes.

File: include/userobjects/VelocityUpdaterBase.h [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+  // setting all of these methods to do nothing since this class will only
+  // provide a method to be called by a ray tracing study
+  virtual void initialSetup() override {}
+  virtual void residualSetup() override {}
+  virtual void jacobianSetup() override {}
+  virtual void meshChanged() override {}
+  virtual void timestepSetup() override {}
+  virtual void initialize() override {}
+  virtual void finalize() override {}
+  virtual void execute() override {}
```
Comment (loganharbour) [inline include/userobjects/VelocityUpdaterBase.h]:
```suggestion
  /**
   * Unused methods
   */
  ///@{
  virtual void initialize() override {}
  virtual void finalize() override {}
  virtual void execute() override {}
  ///@}
```

I think that only the final three need to be overriden. The first ones have default, empty implementations. Also fixes the doxygen comment.

File: include/userobjects/VelocityUpdaterBase.h [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+   * Method for a simple dimenion dependent update of the rays max distance and
+   * velocity based on the dimension of the problem
+   * for a 1D problem only the x value of the velocity stored ray data will be used
+   * for a 2D problem only the x and y values of the velocity store in ray data will be used
+   * for a 3D problem all components of the velocity will be used
+   * @param ray the ray whose velocity we want to set
+   * @param v the new velocity that you want the ray to have
+   * @param dt the time step which you are using to set the maximum distance
+   */
+  virtual void updateVelocity(Ray & ray, const Point v, const Real dt) const;
```
Comment (loganharbour) [inline include/userobjects/VelocityUpdaterBase.h]:
```suggestion
  virtual void updateVelocity(Ray & ray, const Point & v, const Real dt) const;
```

Anything larger than a value should be passed as a reference. 64 bit references are 8 bytes, double precision Points are 8 * 3 = 24 bytes. That is - we save copying 16 bytes by doing this

File: src/userobjects/PICStudyBase.C [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+      // And set the new starting information
+      ray->setStart(start_point, elem);
+      _velocity_updater.updateVelocity(*ray, getVelocity(*ray), _dt);
+    }
+    // Add the rays to be traced
+    moveRaysToBuffer(_banked_rays);
+    _banked_rays.clear();
+  }
+
+  if (!_has_generated)
```
Comment (loganharbour) [inline src/userobjects/PICStudyBase.C]:
```suggestion
  else
```

File: src/userobjects/VelocityUpdaterBase.C [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+  if (dim >= 1)
+    velocity(0) = v(0);
+
+  if (dim >= 2)
+    velocity(1) = v(1);
+
+  if (dim == 3)
+    velocity(2) = v(2);
+
+  // max distance is the
```
Comment (loganharbour) [inline src/userobjects/VelocityUpdaterBase.C]: is the...

File: test/src/userobjects/TestPICStudyUserDefinedRays.C [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+
+    // Claim the rays
+    std::vector<std::shared_ptr<Ray>> claimed_rays;
+    ClaimRays claim_rays(*this, rays, claimed_rays, false);
+    claim_rays.claim();
+    // ...and then add them to be traced
+    moveRaysToBuffer(claimed_rays);
+  }
+
+  PICStudyBase::generateRays();
```
Comment (loganharbour) [inline test/src/userobjects/TestPICStudyUserDefinedRays.C]: So that we don't have this call - what about we create a pure virtual method in the base class `void initializeRays()`, that we then override here instead of `generateRays()`. and then call `initializeRays()` in the base `generateRays()`?

File: test/tests/userobjects/simple_stepping/simple_stepping.i [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+  type = Transient
+  num_steps = 10
+[]
+
+[Problem]
+  solve = false
+[]
+
+[Outputs/rays]
+  type = RayTracingExodus
```
Comment (loganharbour) [inline test/tests/userobjects/simple_stepping/simple_stepping.i:58]: There should be also options to output the ray data here - that way we can test the velocity and (zero) weights too

**Review by loganharbour — CHANGES_REQUESTED** [review]
Design looks good to me. One small design request (the generation part); the rest of the comments are minor

File: test/tests/userobjects/simple_stepping/simple_stepping.i (hunk 23 above)
Comment (gsgall) [reply test/tests/userobjects/simple_stepping/simple_stepping.i:58]: Test input has been updated and regolded to include this change.

File: src/userobjects/PICStudyBase.C (hunk 20 above)
Comment (gsgall) [reply src/userobjects/PICStudyBase.C]: This change was made

File: include/userobjects/VelocityUpdaterBase.h (hunk 19 above)
Comment (gsgall) [reply include/userobjects/VelocityUpdaterBase.h]: This change was made

File: include/userobjects/VelocityUpdaterBase.h (hunk 18 above)
Comment (gsgall) [reply include/userobjects/VelocityUpdaterBase.h]: This change was also made

File: src/userobjects/VelocityUpdaterBase.C (hunk 21 above)
Comment (gsgall) [reply src/userobjects/VelocityUpdaterBase.C]: The commend was updated to be more complete

File: include/userobjects/PICStudyBase.h (hunk 13 above)
Comment (gsgall) [reply include/userobjects/PICStudyBase.h:28]: This member variable was made protected.

File: include/userobjects/PICStudyBase.h [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+   * back to user as a vector to make calculations easier
+   * @param ray the ray
+   */
+  Point getVelocity(const Ray & ray) const;
+
+  /**
+   *  Method that users should override for their custom particle initialization
+   *  This is only called when the study first starts up
+   */
+  virtual void initializeParticles() {}
```
Comment (loganharbour) [inline include/userobjects/PICStudyBase.h]:
```suggestion
  virtual void initializeParticles() = 0;
```

This should be a pure virtual. if you need more info on that concept, let me know.

File: src/userobjects/VelocityUpdaterBase.C [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+  // rays with higher dimensional velocities than the mesh
+  const auto dim = _fe_problem.mesh().dimension();
+
+  if (dim >= 1)
+    velocity(0) = v(0);
+
+  if (dim >= 2)
+    velocity(1) = v(1);
+
+  if (dim == 3)
```
Comment (loganharbour) [inline src/userobjects/VelocityUpdaterBase.C:44]: Screenshot 2023-12-08 at 12 12 02 PM

If you look at the coverage diff, you'll see that this line isn't covered. With that - we should make this a 3D test

File: include/userobjects/PICStudyBase.h [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+  ///@{
+  const RayDataIndex _v_x_index;
+  const RayDataIndex _v_y_index;
+  const RayDataIndex _v_z_index;
+  ///@}
+  /// Ray data fro storing the number of real particles each ray represents
+  const RayDataIndex _weight_index;
+
+  /// Whether or not we've generated rays yet (restartable)
+  bool & _has_generated;
```
Comment (loganharbour) [inline include/userobjects/PICStudyBase.h]: Shouldn't this be private?

File: include/userobjects/PICStudyBase.h (hunk 13 above)
Comment (loganharbour) [reply include/userobjects/PICStudyBase.h:28]: Shouldn't it be private? What derived classes will need to use it? It's not used in a derived class currently.

It's typically better to make things private up front, and then move them to protected if needed.

**Review by loganharbour — CHANGES_REQUESTED** [review]
Just about done

File: src/userobjects/PICStudyBase.C [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+void
+PICStudyBase::generateRays()
+{
+  // We generate rays the first time only, after that we will
+  // pull from the bank and update velocities/max distances
+  if (!_has_generated)
+  {
+    initializeParticles();
```
Comment (loganharbour) [inline src/userobjects/PICStudyBase.C]: Just to get it down on paper (this isn't a review request, just an info request). You do _not_ want to use the same velocity setter that we do in reinit here in the init, correct?

Comment (gsgall) [reply src/userobjects/PICStudyBase.C]: That is correct. The main particle stepping algorithm we will be using for EM, the Boris algorithm is similar to a leap frog method. This requires the first step of particles to be a half time step and then from there we take full time steps.

File: src/userobjects/VelocityUpdaterBase.C (hunk 25 above)
Comment (gsgall) [reply src/userobjects/VelocityUpdaterBase.C:44]: Test has been updated to make it 3D for coverage.

File: include/userobjects/PICStudyBase.h (hunk 26 above)
Comment (gsgall) [reply include/userobjects/PICStudyBase.h]: Yeah you're right. I've updated that to make sure it's private.

File: src/userobjects/PICStudyBase.C [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+//* FENIX is powered by the MOOSE Framework
+//* https://www.mooseframework.inl.gov
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "PICStudyBase.h"
+#include "VelocityUpdaterBase.h"
+
+registerMooseObject("FenixApp", PICStudyBase);
```
Comment (loganharbour) [inline src/userobjects/PICStudyBase.C]: Remove this and you should be able to make the init method a pure virtual.

There's no need to make `PICStudyBase` an object that is usable in input. It's an incomplete base class. It doesn't make sense to instantiate it

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: src/userobjects/PICStudyBase.C (hunk 28 above)
Comment (gsgall) [reply src/userobjects/PICStudyBase.C]: Thank you for the guidance! I just pushed that change.

File: test/src/userobjects/TestPICStudyUserDefinedRays.C (hunk 22 above)
Comment (gsgall) [reply test/src/userobjects/TestPICStudyUserDefinedRays.C]: An `initializeParticles()` method was created for the initial particle creation and a `reinitializeParticles()` method was created to support the re use of particles while avoid the call

File: include/userobjects/PICStudyBase.h (hunk 24 above)
Comment (gsgall) [reply include/userobjects/PICStudyBase.h]: This change was made and the registration of the PICStudyBase was removed in order to accommodate this change.

File: doc/content/source/userobjects/VelocityUpdaterBase.md [hunk 29]
```diff
@@ -0,0 +1,9 @@
+# VelocityUpdaterBase
+
+VelocityUpdaterBase provides the basic method to update a particles direction and maximum distance based on the veloctiy stored in ray data, the current time step and dimension of the mesh. The maximum distance will be set by
+
+\begin{equation}
+  d_\text{max} = \vec{v} \cdot \vec{v} \: dt
+\end{equation}
+
+However, only the components of the velocity data corrisponding to the dimension of the mesh will be used in the calculation. If the mesh is 2D then only the first 2 components of velocity will be used for this calculation. This calculation is peformed with `VelocityUpdaterBase::updateVelocity`. This function takes an argument of the ray that needs its velocity updated, the new velocity of the ray and a time step to be used for the maximum distance calculation to allow flexibility for particle pushing methods.
```
Comment (loganharbour) [inline doc/content/source/userobjects/VelocityUpdaterBase.md:9]:
```suggestion
However, only the components of the velocity data corrisponding to the dimension of the mesh will be used in the calculation. If the mesh is 2D then only the first 2 components of velocity will be used for this calculation. This calculation is peformed with `VelocityUpdaterBase::updateVelocity`. This function takes an argument of the ray that needs its velocity updated, the new velocity of the ray and a time step to be used for the maximum distance calculation to allow flexibility for particle pushing methods.

!syntax parameters /UserObjects/VelocityUpdaterBase

!syntax inputs /UserObjects/VelocityUpdaterBase

!syntax children /UserObjects/VelocityUpdaterBase
```

Forgot to say this - this auto fills the parameters, example input files, and children objects in the documentation

**Review by loganharbour — CHANGES_REQUESTED** [review]
Last change! forgot about doco. code is good

**Review by loganharbour — DISMISSED** [review]

**Comment by loganharbour** [comment]
@cticenhour looks like your review re: headers is still around. I could just dismiss and merge, but let's not do that :-)

File: test/tests/userobjects/simple_stepping/tests [hunk 30]
```diff
@@ -0,0 +1,20 @@
...
+               simple_stepping_rays.e-s0002
+               simple_stepping_rays.e-s0003
+               simple_stepping_rays.e-s0004
+               simple_stepping_rays.e-s0005
+               simple_stepping_rays.e-s0006
+               simple_stepping_rays.e-s0007
+               simple_stepping_rays.e-s0008
+               simple_stepping_rays.e-s0009'
+    allow_test_objects = true
+    requirement = 'PIC Requires the ability to track 3 velocity components but propagate in lower dimensions'
```
Comment (cticenhour) [inline test/tests/userobjects/simple_stepping/tests]:
```suggestion
    requirement = 'The system shall be capable of tracking 3 velocity components within the PIC capability while propagating rays in lower dimensions'
```

**Review by cticenhour — CHANGES_REQUESTED** [review]
All test spec requirements need to be in the form "The system shall....". This comes from our NQA auditors for MOOSE, so we should follow the same standard here. That was the only comment I had, so this PR should be ready to go after this gets fixed up!

**Review by cticenhour — APPROVED** [review]
